rust/src/database/verkle/variants/managed/nodes/sparse_inner.rs

```diff
@@ -0,0 +1,452 @@
+    statistics::node_count::NodeCountVisitor,
+    types::{Key, ToNodeKind},
+};
+
+/// An inner node in a managed Verkle trie.
+// NOTE: Changing the layout of this struct will break backwards compatibility of the
+// serialization format.
+#[derive(Debug, Clone, PartialEq, Eq, FromBytes, IntoBytes, Immutable, Unaligned)]
+#[repr(C)]
+pub struct SparseInnerNode<const N: usize> {
+    pub children: [VerkleIdWithIndex; N],
+    pub commitment: VerkleCommitment,
+}
+
+impl<const N: usize> SparseInnerNode<N> {
+    /// Creates a sparse inner node from existing children and commitment.
+    /// Returns an error if there are more than N non-zero children.
+    pub fn from_existing(
+        children: &[VerkleIdWithIndex],
+        commitment: &VerkleCommitment,
+    ) -> BTResult<Self, Error> {
+        let mut inner = SparseInnerNode {
+            commitment: *commitment,
+            ..Default::default()
+        };
+
+        // Insert values from previous leaf using get_slot_for to ensure no duplicate indices.
+        for vwi in children {
+            if vwi.item == VerkleNodeId::default() {
+                continue;
+            }
+            let slot = VerkleIdWithIndex::get_slot_for(&inner.children, vwi.index).ok_or(
+                Error::CorruptedState(format!(
+                    "too many non-zero IDs to fit into sparse inner of size {N}"
+                )),
+            )?;
+            inner.children[slot] = *vwi;
+        }
+
+        Ok(inner)
+    }
+
+    /// Returns the children of this inner node as commitment input.
+    // TODO: This should not have to pass 256 values: https://github.com/0xsoniclabs/sonic-admin/issues/384
+    pub fn get_commitment_input(&self) -> BTResult<VerkleCommitmentInput, Error> {
+        let mut values = [VerkleNodeId::default(); 256];
+        for VerkleIdWithIndex { index, item: value } in &self.children {
+            values[*index as usize] = *value;
+        }
+        Ok(VerkleCommitmentInput::Inner(values))
+    }
+}
+
+impl<const N: usize> Default for SparseInnerNode<N> {
+    fn default() -> Self {
+        let mut children = [VerkleIdWithIndex::default(); N];
+        children.iter_mut().enumerate().for_each(|(i, v)| {
+            v.index = i as u8;
+        });
+
+        SparseInnerNode {
+            children,
+            commitment: VerkleCommitment::default(),
+        }
+    }
+}
+
+impl<const N: usize> ManagedTrieNode for SparseInnerNode<N> {
+    type Union = VerkleNode;
+    type Id = VerkleNodeId;
+    type Commitment = VerkleCommitment;
+
+    fn lookup(&self, key: &Key, depth: u8) -> BTResult<LookupResult<Self::Id>, Error> {
+        let slot = VerkleIdWithIndex::get_slot_for(&self.children, key[depth as usize]);
+        match slot {
+            Some(slot) if self.children[slot].index == key[depth as usize] => {
+                Ok(LookupResult::Node(self.children[slot].item))
+            }
+            _ => Ok(LookupResult::Node(VerkleNodeId::default())),
```

Comment:
```suggestion
            Some(slot) => {
                Ok(LookupResult::Node(self.children[slot].item))
            }
            None => Ok(LookupResult::Node(VerkleNodeId::default())),
```

There's no need to check if the index matches, as the item would be `VerkleNodeId::default()` anyway.